File: packages/types/src/llm/request.ts
/**
 * LLM Request Type Definition
 */

import type { ID } from "../common.js";
import type { Message } from "../message/index.js";
import type { ToolSchema } from "../tool/index.js";
import type { ToolCallFormatConfig } from "./tool-call-format.js";
import type { ToolCallProtocolViolationPolicy } from "./protocol-config.js";

/**
 * Dead loop detection configuration for LLM requests
 */
export interface DeadLoopDetectionConfig {
  /** Enable dead loop detection (default: true) */
  enabled?: boolean;
  /** Checkpoint thresholds (character count) */
  checkpoints?: number[];
  /** Short sequence detection window size (characters) */
  shortSequenceWindow?: number;
  /** Minimum repeat unit length (characters) */
  minRepeatUnitLength?: number;
  /** Minimum repeat count */
  minRepeatCount?: number;
  /** Minimum period elements (blocks/lines) */
  minPeriodElements?: number;
  /** Maximum period length */
  maxPeriodLength?: number;
}

/**
 * LLM Request Type
 */
export interface LLMRequest {
  /** Referenced LLM Profile ID (optional, default configuration used if not provided) */
  profileId?: ID;
  /** message array */
  messages: Message[];
  /** Request Parameters object (overrides parameters in Profile) */
  parameters?: Record<string, unknown>;
  /** Definition of available tools */
  tools?: ToolSchema[];
  /**
   * Tool call format configuration override.
   *
   * @deprecated This field is currently not used in the execution path.
   * Use `lockedToolCallFormat` instead, which is set by the executor at execution start
   * and provides protocol locking and enforcement. The `toolCallFormat` field on
   * `LLMProfile` is the source of truth for the profile-level format.
   * If you need per-request overrides, set the profile's toolCallFormat or use
   * the AgentLoopDefinition's toolCallFormat.
   */
  toolCallFormat?: ToolCallFormatConfig;

  /**
   * Locked tool call format for this execution.
   * Set by the executor at execution start.
   * Overrides profile-level toolCallFormat.
   * The LLMClient must use this format regardless of profile changes.
   */
  lockedToolCallFormat?: ToolCallFormatConfig;

  /**
   * Execution ID for tracing and correlation.
   * Set by the executor to correlate log entries and metrics
   * across the full LLM call chain.
   */
  executionId?: string;

  /**
   * Per-request protocol violation policy override.
   * When set, overrides the global default policy for this request only.
   * Resolution order: request-level > agent-level > global default.
   */
  violationPolicy?: ToolCallProtocolViolationPolicy;

  /** Streaming or not */
  stream?: boolean;
  /** AbortSignal for interrupt requests */
  signal?: AbortSignal;
  /** Dead loop detection configuration */
  deadLoopDetection?: DeadLoopDetectionConfig;
}

File: crates/foundation/wf-types/src/llm/request.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DeadLoopDetectionConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
    /// Character-length checkpoints at which detection runs
    /// (default: [500, 1000, 2000]).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub checkpoints: Option<Vec<u32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub short_sequence_window: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_repeat_unit_length: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_repeat_count: Option<u32>,
    /// Minimum number of consecutive equal elements (paragraphs / normalized
    /// lines) that must repeat to be flagged as a loop (default: 6).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_period_elements: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_period_length: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ToolCallProtocolViolationPolicy {
    Ignore,
    Warn,
    Fail,
    AutoConvert,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LlmRequest {
    pub profile_id: String,
    pub messages: Vec<super::super::message::Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation: Option<super::generation::LlmGenerationParams>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<super::super::tool::Tool>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_format: Option<super::ToolCallFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub locked_tool_call_format: Option<super::tool_call_format::ToolCallFormatConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub violation_policy: Option<ToolCallProtocolViolationPolicy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dead_loop_detection: Option<DeadLoopDetectionConfig>,
    /// Set by the gateway when a locked tool call format conflict was resolved
    /// via the `auto_convert` policy; observed by formatters for observability.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_auto_converted: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ChatRequest {
    pub model: String,
    pub messages: Vec<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<serde_json::Value>>,
}
